core-api/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,263 @@
+pub struct MetricsAttributesOptions {
+    attributes: Vec<MetricKeyValue>,
+}
+impl MetricsAttributesOptions {
+    pub fn extend(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        self.attributes.extend(new_kvs.into_iter())
+    }
+}
+
+/// A K/V pair that can be used to label a specific recording of a metric
+#[derive(Clone, Debug)]
+pub struct MetricKeyValue {
+    key: String,
+    value: MetricValue,
+}
+impl MetricKeyValue {
+    pub fn new(key: impl Into<String>, value: impl Into<MetricValue>) -> Self {
+        Self {
+            key: key.into(),
+            value: value.into(),
+        }
+    }
+}
+
+/// Values metric labels may assume
+#[derive(Clone, Debug, derive_more::From)]
+pub enum MetricValue {
+    String(String),
+    Int(i64),
+    Float(f64),
+    Bool(bool),
+    // can add array if needed
+}
+impl From<&'static str> for MetricValue {
+    fn from(value: &'static str) -> Self {
+        MetricValue::String(value.to_string())
+    }
+}
+
+/// Histogram buckets
+#[derive(Clone, Debug)]
+pub struct Buckets {
+    buckets: Vec<f64>,
+}
+
+pub trait Counter: Send + Sync {
+    fn add(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub trait Histogram: Send + Sync {
+    // When referring to durations, this value is in millis
+    fn record(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub trait Gauge: Send + Sync {
+    // When referring to durations, this value is in millis
+    fn record(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub struct NoOpInstrument;
+impl Counter for NoOpInstrument {
+    fn add(&self, _: u64, _: &MetricAttributes) {}
+}
+impl Histogram for NoOpInstrument {
+    fn record(&self, _: u64, _: &MetricAttributes) {}
+}
+impl Gauge for NoOpInstrument {
+    fn record(&self, _: u64, _: &MetricAttributes) {}
+}
+
+#[cfg(feature = "otel_impls")]
+mod otel_impls {
+    use super::*;
+    use opentelemetry::{metrics, metrics::Meter, KeyValue};
+
+    impl From<MetricKeyValue> for KeyValue {
+        fn from(kv: MetricKeyValue) -> Self {
+            KeyValue::new(kv.key, kv.value)
+        }
+    }
+
+    impl From<MetricValue> for opentelemetry::Value {
+        fn from(mv: MetricValue) -> Self {
+            match mv {
+                MetricValue::String(s) => opentelemetry::Value::String(s.into()),
+                MetricValue::Int(i) => opentelemetry::Value::I64(i),
+                MetricValue::Float(f) => opentelemetry::Value::F64(f),
+                MetricValue::Bool(b) => opentelemetry::Value::Bool(b),
+            }
+        }
+    }
+
+    impl CoreMeter for Meter {
+        fn new_attributes(&self, attribs: MetricsAttributesOptions) -> MetricAttributes {
+            MetricAttributes::OTel {
+                kvs: Arc::new(attribs.attributes.into_iter().map(KeyValue::from).collect()),
+                // TODO: Pass in? Hard to make work with trait
+                ctx: opentelemetry::Context::current(),
```

Comment:
Just documenting that users can use thread-locals or thread contexts or whatever is enough I think.